firstpartydata/first_party_data_test.go

```diff
@@ -0,0 +1,836 @@
+  				  }
+  				},
+  				"user": {
+  				  "id": "appId",
+				  "ext": {}
+  				},
+  				"tmax": 5000,
+  				"source": {
+  				  "tid": "ad839de0-5ae6-40bb-92b2-af8bad6439b3"
+  				}
+			}`),
+			expectedFpd: map[string][]byte{
+				"app":  {},
+				"user": []byte(`{"someuserfpd": "userfpdDataTest"}`),
+				"site": {},
+			},
+		},
+		{
+			description: "No FPD present in req",
+			input: []byte(`{
+  				"id": "bid_id",
+  				"site": {
+  				  "id":"reqSiteId",
+  				  "page": "http://www.foobar.com/1234.html",
+  				  "publisher": {
+  				    "id": "1"
+  				  }
+  				},
+  				"app": {
+  				  "id": "appId",
+  				  "ext": {
+                  }
+  				},
+			  	"user": {
+  				  "id": "userId",
+				  "ext": {}
+  				},
+  				"tmax": 5000,
+  				"source": {
+  				  "tid": "ad839de0-5ae6-40bb-92b2-af8bad6439b3"
+  				}
+			}`),
+			output: []byte(`{
+  				"id": "bid_id",
+  				"site": {
+  				  "id":"reqSiteId",
+  				  "page": "http://www.foobar.com/1234.html",
+  				  "publisher": {
+  				    "id": "1"
+  				  }
+  				},
+  				"app": {
+  				  "id": "appId",
+				  "ext": {}
+  				},
+				"user": {
+  				  "id": "userId",
+				  "ext": {}
+  				},
+  				"tmax": 5000,
+  				"source": {
+  				  "tid": "ad839de0-5ae6-40bb-92b2-af8bad6439b3"
+  				}
+			}`),
+			expectedFpd: map[string][]byte{
+				"app":  {},
+				"user": {},
+				"site": {},
+			},
+		},
+		{
+			description: "Site FPD only present",
+			input: []byte(`{
+  				"id": "bid_id",
+  				"site": {
+  				  "id":"reqSiteId",
+  				  "page": "http://www.foobar.com/1234.html",
+  				  "publisher": {
+  				    "id": "1"
+  				  },
+				  "ext": {
+					"data": {"someappfpd": true},
+					"amp": 1
+                  }
+  				},
+  				"app": {
+  				  "id": "appId"
+  				},
+  				"tmax": 5000,
+  				"source": {
+  				  "tid": "ad839de0-5ae6-40bb-92b2-af8bad6439b3"
+  				}
+			}`),
+			output: []byte(`{
+  				"id": "bid_id",
+  				"site": {
+  				  "id":"reqSiteId",
+  				  "page": "http://www.foobar.com/1234.html",
+  				  "publisher": {
+  				    "id": "1"
+  				  },
+                 "ext": {
+					"amp": 1
+                  }
+  				},
+  				"app": {
+  				  "id": "appId"
+  				},
+  				"tmax": 5000,
+  				"source": {
+  				  "tid": "ad839de0-5ae6-40bb-92b2-af8bad6439b3"
+  				}
+			}`),
+			expectedFpd: map[string][]byte{
+				"app":  {},
+				"user": {},
+				"site": []byte(`{"someappfpd": true}`),
+			},
+		},
+	}
+	for _, test := range testCases {
+		var inputTestReq openrtb_ext.RequestWrapper
+		err := json.Unmarshal(test.input, &inputTestReq)
+		assert.NoError(t, err, "Error should be nil")
+
+		fpd, err := ExtractGlobalFPD(&inputTestReq)
+		inputTestReq.RebuildRequest()
+		assert.NoError(t, err, "Error should be nil")
+
+		var outputTestReq openrtb_ext.RequestWrapper
+		err = json.Unmarshal(test.output, &outputTestReq)
+		assert.NoError(t, err, "Error should be nil")
+
+		if fpd[userKey] != nil {
```

Comment:
What happens if the global FPD returned by `ExtractGlobalFPD` doesn't have a user key even though we expected it? In this case this condition will be skipped and the test will silently pass even though it shouldn't have. Also, I am not sure if we even need this check. Shouldn't `assert.Equal` take care of this?

Reply:
Yes, fixed, good point!